Add local community detection algorithms

michitux opened the pull request:
This is the code that I have developed for our paper on local community detection:

Hamann, Michael; Röhrs, Eike; Wagner, Dorothea. 2017. "[Local Community Detection based on Small Cliques](http://www.mdpi.com/1999-4893/10/3/90)." Algorithms 10, no. 3: 90.

The scripts that are based on this code can be found at https://github.com/kit-algo/LCD-cliques-experiments

Apart from adding many local community detection algorithms, this also adds a class for evaluating the results of local community detection algorithms with respect to a ground truth cover. Further, some helper classes have been added.

I have rebased this code on the current NetworKit master branch and applied clang-format to all new files.

All algorithms try to be really local in the sense that they (should) never allocate memory proportional to the graph size (apart from `RandomBFS` maybe, which is just a baseline). To detect many communities, it would definitely be faster to instead use global data structures, and if the graph is not too large compared to a found community, this would most likely also speed up even finding a single community.

michitux commented:
I have extracted the common code into a separate helper class and I have tried to adapt the code even more to current NetworKit standards so this is ready for review now. I have created two helper classes for different algorithms. One is very closely tied to some of the algorithms and currently just lives as a `.hpp` (as it is a class template) in the implementation folder. The other is slightly more versatile and currently lives in `structures`, but I could also see reasons to also move it into the `scd` implementations directory. Any opinions?

This code also introduces a small interface change, I renamed the method `expandSeed` that existed in `GCE` and `PageRankNibble` to `expandOneCommunity` as I found the latter clearer in the case multiple seed nodes are specified (which is a new feature). The old method had not been exported to Python. I could re-add the old method as a deprecated alias if this is necessary for the deprecation policy.

avdgrinten commented:
Thank you for the PR Michael. Yes, adding a deprecated alias would be nice. I should be able to review the PR during this week.

angriman commented on include/networkit/scd/CliqueDetect.hpp:
```diff
@@ -0,0 +1,53 @@
+ *
+ * The algorithm can handle weighted graphs.
+ */
+class CliqueDetect : public NetworKit::SelectiveCommunityDetector {
```
I guess that writing `NetworKit::` is redundant here since we are already in the NetworKit namespace.

michitux replied:
Fixed in 56c52c5639441be8bda89fbac450c6c1689c709b

angriman commented on include/networkit/scd/CliqueDetect.hpp:
```diff
@@ -0,0 +1,53 @@
+
+protected:
+    template <typename F>
+    std::vector<node> getMaximumWeightClique(const std::vector<node> &nodes, F is_seed) const;
```
Use lowerCamelCase for `is_seed`.

michitux replied:
Fixed in 35b4f77ef9b92eb98e5c2c0e19ccfe6a7aa3620e

angriman commented on include/networkit/scd/CliqueDetect.hpp:
```diff
@@ -0,0 +1,53 @@
+    template <typename F>
+    std::vector<node> getMaximumWeightClique(const std::vector<node> &nodes, F is_seed) const;
+
+    Graph createSubgraphFromNodes(const std::vector<node> &nodes) const;
```
Isn't this method already implemented in `GraphTools::subgraphFromNodes`?

michitux replied:
No, the method in `GraphTools` is linear in the size of the input graph and creates a graph with the same node ids as the input graph (i.e., O(n) nodes), while this method creates a graph with compact node ids and is linear in the volume of the given nodes. It could make sense to offer this method with compact node ids in `GraphTools`, though.

angriman replied:
In that case I'd suggest to move this method to `GraphTools` and provide some unit tests.

angriman commented on include/networkit/scd/LFMLocal.hpp:
```diff
@@ -0,0 +1,54 @@
+ * give a community containing the whole graph while values larger
+ * than 2 recover the smallest communities.
+ */
+class LFMLocal : public NetworKit::SelectiveCommunityDetector {
```
`NetworKit::` is redundant here.

michitux replied:
Fixed in 56c52c5639441be8bda89fbac450c6c1689c709b

angriman commented on include/networkit/scd/LocalT.hpp:
```diff
@@ -0,0 +1,40 @@
+ * In 2014 IEEE/ACM International Conference on Advances in Social Networks Analysis and Mining
+ * (ASONAM) (pp. 108–112). https://doi.org/10.1109/ASONAM.2014.6921568
+ */
+class LocalT : public NetworKit::SelectiveCommunityDetector {
```
As above.

michitux replied:
Fixed in 56c52c5639441be8bda89fbac450c6c1689c709b

angriman commented on include/networkit/scd/SCDGroundTruthComparison.hpp:
```diff
@@ -0,0 +1,121 @@
+ * For each score, the range of values is between 0 and 1, where 0 is the worst and 1 the best
+ * score.
+ */
+class SCDGroundTruthComparison : public Algorithm {
```
Make this class `final`?

angriman commented on include/networkit/scd/SelectiveCommunityDetector.hpp:
```diff
@@ -16,22 +16,57 @@
+     * @param seeds The seeds for the community.
+     * @return The found community as set of nodes.
+     */
+    virtual std::set<node> expandOneCommunity(const std::set<node>& seeds) = 0;
```
Instead of a `std::set`, what about a range of nodes?

michitux replied:
Accepting a range of nodes means that this method must be a template method, but template methods cannot be virtual. Having this method as a virtual method allows accepting any selective community detector and executing it from a set of seed nodes.

angriman replied:
Ah that's right. In the implementation of this method the order in which you process the seeds does not seem to be relevant. Would that be better to use a `std::unordered_set` for `seeds`?

michitux replied:
I wanted to keep this consistent with the return type of the already existing local community detection algorithms, which is `std::set`, as my primary application of multiple seed nodes is to use the result of a local community detection algorithm (in particular `CliqueDetect`) as seed for another local community detection algorithm. I would also be open for a change to `std::vector` as it should have much less overhead compared to `std::set` or `std::unordered_set`.

angriman commented on include/networkit/scd/SetConductance.hpp:
```diff
@@ -0,0 +1,47 @@
+     * @param G The graph
+     * @param community The set of nodes to examine.
+     */
+    SetConductance(const Graph &G, const std::set<node> &community);
```
Range instead of `std::set`?

angriman commented on networkit/cpp/scd/CliqueDetect.cpp:
```diff
@@ -0,0 +1,158 @@
+    std::set<node> result;
+    result.insert(s);
+
+    std::vector<node> sn;
+    sn.reserve(G->degree(s));
+    G->forNeighborsOf(s, [&](node v) { sn.push_back(v); });
```
What about `std::vector<node> sn(G->neighborRange(s).begin(), G->neighborRange(s).end())`

michitux replied:
The code predates the introduction of `neighborRange`, I agree that it is much nicer and also changed the remainder of this algorithm to use neighbor ranges. I also discovered a bug in the implementation for weighted graphs that I fixed.

angriman commented on networkit/cpp/scd/CliqueDetect.cpp:
```diff
@@ -0,0 +1,158 @@
+
+template <typename F>
+std::vector<node> CliqueDetect::getMaximumWeightClique(const std::vector<node> &nodes,
+                                                       F is_seed) const {
```
Use lowerCamelCase for `is_seed`

michitux replied:
Fixed in 35b4f77ef9b92eb98e5c2c0e19ccfe6a7aa3620e

angriman reviewed (commented):
Since this PR is quite long I reviewed just a portion of it for now. I have two general comments about the style:
- The new files should include the line `// networkit-format`
- As written in the review comments, there are several instances of `class MyClass : public NetworKit::BaseClass`. I think that `NetworKit::` is redundant here since we are already in the NetworKit namespace, I think it should be removed.

angriman commented on networkit/cpp/scd/LocalT.cpp:
```diff
@@ -0,0 +1,191 @@
+    return result;
+}
+
+} /* namespace NetworKit */
```
In this file many variables use snake_case.

michitux replied:
I changed all variables to use lowerCamelCase.

angriman reviewed (commented):
I added some new comments. Another general style-related issue that I see in this PR is that some files use a mixture of snake_case and camelCase.

michitux commented on networkit/cpp/scd/LocalDegreeDirectedGraph.hpp:
```diff
@@ -0,0 +1,281 @@
+};
+
+} // namespace NetworKit
+#endif
```
I changed all variables to use lowerCamelCase.

michitux commented on networkit/cpp/scd/LocalDegreeDirectedGraph.hpp:
```diff
@@ -0,0 +1,281 @@
+        /** first index in head where neighbors of the node are stored */
+        std::size_t first_head;
+        /** index after the last index in head where neighbors of the node are stored */
+        std::size_t last_head;
```
That seems to be correct, removed.

michitux commented on networkit/cpp/scd/LocalDegreeDirectedGraph.hpp:
```diff
@@ -0,0 +1,281 @@
+
+    std::vector<node> current_local_neighbor_ids;
+
+    const NetworKit::Graph &G;
```
Removed `NetworKit::`.

angriman commented on networkit/cpp/scd/LocalTightnessExpansion.cpp:
```diff
@@ -0,0 +1,297 @@
+    public:
+        Compare(const std::vector<double> &similarity) : similarity(similarity) {}
+
+        bool operator()(node u, node v) { return similarity[u] > similarity[v]; }
```
This function can be `const`

angriman commented on networkit/cpp/scd/LocalTightnessExpansion.cpp:
```diff
@@ -0,0 +1,297 @@
+                        nodeExternalSimilarity[lu] += scoreUv;
+                    }
+
+                    if (shell.contains(lv)) {
+                        shell.update(lv);
+                    } else {
+                        shell.push(lv);
+                    }
```
You can simplify this code by simply using `shell.update(lv)`, `update` already implements this check.

angriman commented on networkit/cpp/scd/LocalTightnessExpansion.cpp:
```diff
@@ -0,0 +1,297 @@
+    }
+}
+
+} /* namespace NetworKit */
```
Some debugging checks are quite long and make the code harder to follow. If it does not require too much work, I would suggest to implement them in separate functions.

angriman commented on include/networkit/scd/SCDGroundTruthComparison.hpp:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Get the precision of every found community.
+     *
+     * @return A map between seed node and the precision of the seed's community.
+     */
+    std::map<index, double> getIndividualPrecision();
+
+    /**
+     * Get the recall of every found community.
+     *
+     * @return A map between seed node and the recall of the seed's community.
+     */
+    std::map<index, double> getIndividualRecall();
+
+    /**
+     * Get the F1 score of every found community.
+     *
+     * @return A map between seed node and the F1 score of the seed's community.
+     */
+    std::map<index, double> getIndividualF1();
+
+    /**
+     * Get the (unweighted) average of the jaccard indices of every found community.
+     */
+    double getAverageJaccard();
+
+    /**
+     * Get the (unweighted) average of the F1 score of every found community.
+     */
+    double getAverageF1();
+
+    /**
+     * Get the (unweighted) average of the precision of every found community.
+     */
+    double getAveragePrecision();
+
+    /**
+     * Get the (unweighted) average of the recall of every found community.
+     */
+    double getAverageRecall();
```
These functions can be `const`.

angriman replied:
Also, why the ones returning a `std::map` do not return a const ref?

angriman commented on networkit/cpp/scd/TCE.cpp:
```diff
@@ -0,0 +1,306 @@
+    public:
+        Compare(const std::vector<double> &similarity) : similarity(similarity) {}
+
+        bool operator()(node u, node v) { return similarity[u] > similarity[v]; }
```
This method can be `const`.

angriman commented on networkit/cpp/scd/TCE.cpp:
```diff
@@ -0,0 +1,306 @@
+                                            uNeighbors, useJaccard));
+#endif
+                nodeScore[v] += scoreUv;
+                if (shell.contains(v)) {
+                    shell.update(v);
+                } else {
+                    shell.push(v);
+                }
```
As before, simplify with just `shell.update(v);`.

angriman commented on networkit/cpp/scd/TwoPhaseL.cpp:
```diff
@@ -0,0 +1,114 @@
+            currentLin = 2.0 * com.internalEdgeWeight() / com.size();
+            currentLex = com.cut() / com.boundarySize();
+            currentL = currentLin / currentLex;
+            // TRACE("community: ", community);
```
Remove this line if unused.

angriman commented on networkit/cpp/scd/test/SelectiveCDGTest.cpp:
```diff
@@ -50,20 +73,172 @@ TEST_F(SCDGTest2, testSCD) {
+
+    // The cluster should only grow
+    for (node u : cluster) {
+        EXPECT_TRUE(cluster2.find(u) != cluster2.end());
```
`EXPECT_NE(cluster2.find(u), cluster2.end());`

angriman commented on networkit/cpp/structures/LocalCommunity.cpp:
```diff
@@ -0,0 +1,317 @@
+#include <networkit/structures/LocalCommunity.hpp>
```
`// networkit-format`

michitux commented on networkit/cpp/scd/LocalDegreeDirectedGraph.hpp:
```diff
@@ -0,0 +1,281 @@
+                }
+            }
+
+            headInfo.emplace_back(LocalNode{myBegin, myEnd});
```
This does not compile because `LocalNode` does not have any corresponding constructor with two arguments.

angriman commented on include/networkit/auxiliary/UniformRandomSelector.hpp:
```diff
@@ -0,0 +1,52 @@
+ * can be reset when a new class of elements is found
+ * that shall be used instead.
+ */
+class UniformRandomSelector {
```
Since the elements are discovered incrementally, shouldn't this class include `Incremental` in its name? Just `UniformRandomSelector` looks too generic. Also, this file can be formatted.

michitux replied:
Renamed to `IncrementalUniformRandomSelector` and added `networkit-format`.

angriman commented:
This PR looks fine to me now, I left just one minor comment. @michitux are you planning do make other changes or do you think that this is ready to be merged?

michitux commented:
I am still planning to move `subgraphFromNodes` to `GraphTools`, though this is kind of waiting for #651 and the parameters are kind of incompatible as the `GraphTools` method expects an `unordered_set`, but I need to know the mapping from the compacted node ids to the original node ids. There are also some other unresolved comments, e.g., on using a range for `SetConductance` as input.

angriman commented:
@michitux ping me when this is ready to be reviewed. Before that please rebase on master.

michitux commented:
@angriman I have addressed the last comment (renaming of `UniformRandomSelector` and rebased on master. Further, I have combined some commits to have a shorter history and fewer bug-fixing commits. Moving the debugging code into separate functions would complicate the code quite a lot, therefore I would suggest to leave it the way it is. I have also added some checks for directed graphs - which are unsupported - and fixed counting self-loops for conductance. The `subgraphFromNodes`-change depends on #726, I have proposed an implementation at Relux-the-Relux/networkit#1. Unless you object I would suggest to merge this PR as-is and change this later after #726 has been merged. If you think #726 can be merged soon, we can of course also wait and I can rebase again and fix this.

michitux commented:
I have now wrapped all asserts that do more than just comparing two already calculated values in `#ifdef NETWORKIT_SANITY_CHECKS` such that these algorithms are not slowed down if asserts should be enabled in release builds as proposed in #769.

angriman commented on include/networkit/auxiliary/IncrementalUniformRandomSelector.hpp:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @return If the new element shall be accepted/kept.
+     */
+    bool add_element() {
```
Coding style

michitux commented on include/networkit/scd/ApproximatePageRank.hpp:
```diff
@@ -19,25 +21,25 @@ namespace NetworKit {
+     * @return Approximate PageRank vector from @a seeds with parameters
      *         specified in the constructor.
      */
-    std::vector<std::pair<node, double>> run(node seed);
```
I have restored the old method as a wrapper, I think we can just keep it.

michitux commented on include/networkit/scd/RandomBFS.hpp:
```diff
@@ -0,0 +1,34 @@
+ * finds a community around a seed node with a given size using a prefix of a
+ * BFS order that is selected at random among all possible prefixes.
+ */
+class RandomBFS : public SelectiveCommunityDetector {
```
I have added a test case that actually found a bug that I've fixed now - tests are really a good idea, thanks.

michitux commented on networkit/cpp/structures/LocalCommunity.cpp:
```diff
@@ -0,0 +1,330 @@
+    return result;
+}
+
+template <bool a, bool b, bool c>
```
Probably because they are unused. I have renamed them.
